frontend_tests/node_tests/popover_menus_data.js

```diff
@@ -0,0 +1,248 @@
+    assert.equal(response.narrowed, false);
+    assert.equal(response.should_display_delete_option, true);
+    assert.equal(response.should_display_read_receipts_option, true);
+    assert.equal(response.should_display_reminder_option, false);
+    assert.equal(response.should_display_quote_and_reply, true);
+});
+
+test("not_my_message_view_actions", () => {
+    page_params.is_spectator = false;
+    page_params.realm_allow_message_editing = true;
+    page_params.realm_message_content_edit_limit_seconds = null;
+    page_params.realm_allow_edit_history = true;
+    page_params.realm_message_content_delete_limit_seconds = null;
+    page_params.realm_delete_own_message_policy = 1;
+    page_params.realm_enable_read_receipts = true;
+
+    // Get message that is only viewable
+
+    const list = init_message_list();
+    message_lists.set_current(list);
+
+    const messages = [
+        {
+            id: 1,
+            sender_id: mike.user_id,
+            is_hidden: false,
+            sent_by_me: false,
+            locally_echoed: false,
+            is_stream: true,
+            stream_id: 1,
+            unread: false,
+            edit_history: [
+                {
+                    prev_content: "Previous content",
+                    prev_stream: 0,
+                    prev_topic: "Previous topic",
+                },
+            ],
+        },
+    ];
+
+    add_message_with_view(list, messages);
+
+    const response = popover_menus_data.get_actions_popover_content_context(1);
+
+    assert.equal(response.view_source_menu_item, $t({defaultMessage: "View message source"}));
+    assert.equal(response.editability_menu_item, undefined);
+    assert.equal(response.move_message_menu_item, undefined);
+});
+
+test("not_my_message_view_source_and_move", () => {
+    page_params.is_spectator = false;
+    page_params.realm_allow_message_editing = true;
+    page_params.realm_message_content_edit_limit_seconds = null;
+    page_params.realm_allow_edit_history = true;
+    page_params.realm_message_content_delete_limit_seconds = null;
+    page_params.realm_delete_own_message_policy = 1;
+    page_params.realm_enable_read_receipts = true;
+    page_params.realm_edit_topic_policy = 5;
+
+    // Get message that is movable with viewable source
+
+    const list = init_message_list();
+    message_lists.set_current(list);
+
+    const messages = [
+        {
+            id: 1,
+            sender_id: mike.user_id,
+            is_hidden: false,
+            sent_by_me: false,
+            locally_echoed: false,
+            is_stream: true,
+            stream_id: 1,
+            type: "stream",
+            unread: false,
+            topic: compose.empty_topic_placeholder(),
+            edit_history: [
+                {
+                    prev_content: "Previous content",
+                    prev_stream: 0,
+                    prev_topic: "Previous topic",
+                },
+            ],
+        },
+    ];
```

Comment:
The message objects in particular are a lot of copy-pasted content. Maybe that's correct, but you should write comments for each one explaining the choices you're making for the message, so that the intent of the test would be readable by a friend who hasn't read the `popover_menus` codebase.

https://zulip.readthedocs.io/en/latest/testing/philosophy.html may be helpful reading.